### DISCUSSION TITLE
New Customer Group

### DISCUSSION THREAD

**Discussion by ali-apadana** [bagisto/bagisto#8583] [General]
Hello everyone
I have a question
we have shop on bagisto and we have more than 100,000 products and we added a new customer group.
Should we update all products one by one to rebuild product_price_indices table?
It's not possible!

**Comment by ali-apadana** [comment]
I have 430,000 records in product_price_indices and now I want to create promotion I'm getting "Gateway Timeout".

**Reply by devansh-webkul** [reply]
We have fixed some issues in the master. Kindly take a pull from the master. In the beta version, there is a long processing task which we have moved in the background. Kindly setup a proper queue driver instead of sync.

**Reply by ali-apadana** [reply]
Thanks. what should I set to? In config/queue.php I have These

`'default' => env('QUEUE_DRIVER', 'sync'),`

**Reply by ali-apadana** [reply]
I changed it to "database" without success

**Comment by ali-apadana** [comment]
please tell me what should i do step by step.
I read codes. there is no chunk. how can i use queue?

**Comment by ali-apadana** [comment]
I traced. In "CatalogRule/src/Helpers/CatalogRuleProduct.php" In method "getMatchingProductIds" It's FREEZING and causes Gateway Timeout.
(I have just 150,000 rows In products table)

**Comment by jitendra-webkul** [comment]
Please go through the "Laravel Queue" documentation.

https://laravel.com/docs/10.x/queues

**Reply by ali-apadana** [reply]
Thanks alot.
But Now problem is before insert/update DB. At "CatalogRule/src/Helpers/CatalogRuleProduct.php" In method "getMatchingProductIds". It wants too get matched products.

**Reply by jitendra-webkul** [reply]
You can use the timeout param to 0 that will work without timeout error

`php artisan queue:listen --timeout=0`

OR

`php artisan queue:work --timeout=0`

For more info reference the following link:

https://laravel.com/docs/10.x/queues#worker-timeouts

**Reply by ali-apadana** [reply]
So It might takes long time. Am I Right?

**Reply by jitendra-webkul** [reply]
Depends on the catalog size that why we moved this to background process

**Reply by ali-apadana** [reply]
Yes, and I'm testing on 30,000 products (plus 120,000 variants)

And You are saying you moved to background. but as I see your codes It's not!
Can you show me piece of code?

As I see you are fetching all products in "getMatchingProductIds" method in "Helpers/CatalogRuleProduct.php". And parse them one by one.

**Reply by ali-apadana** [reply]
wait a minute. You mean of "we moved this to background process" Is process in backend (admin side)?
I thought you're saying process in background without browser (or admin) wait until end. like cronjob

**Reply by jitendra-webkul** [reply]
Indexing won't happen during the Updation or Creation of a catalog rule. We are processing catalog rule indexing in the background. 

This means, that once you create or update a catalog rule after that action we are dispatching a Job that job gets queued in the configured queue driver (Ex. Redis, Database etc.). 

Then you need to set up the supervisor in order to process the queued jobs in background that will do the indexing in the background.

https://laravel.com/docs/10.x/queues#supervisor-configuration

**Reply by ali-apadana** [reply]
Thanks alot.
I sat up & configured supervisor.
I ran "sudo supervisorctl start laravel-worker:*"
I ran "php artisan queue:table", "php artisan migrate​"
I changed .env to "QUEUE_DRIVER=database​" ,  "QUEUE_CONNECTION=database"

But it is not done yet :(

Is there anything else I should do?

Thanks for help

**Reply by jitendra-webkul** [reply]
Which version are you using?

If you are using 2.0.0-BETA-1 then it won't work as this version doesn't support queue yet. It's recently added in the master branch.

Please take a pull from the master branch and try again

**Reply by ali-apadana** [reply]
In home directory I ran "git branch" I Got "fatal: not a git repository (or any of the parent directories): .git"
I've downloded master and upload to my server.
Am I wrong?

**Reply by devansh-webkul** [reply]
Use git clone command. If you download zip then git tracking wont be available.

**Comment by ali-apadana** [comment]
Thanks
Another thing is problem with sort price.
when you are saving product you set elastic properties by "price_CUSTOMERGROUPID" Key. and when you want to search, search by "price" key. and this cause of error "No mapping found for [price] in order to sort on"
